selfservice/flow/login/handler.go

```diff
@@ -407,6 +407,19 @@ func (h *Handler) createBrowserLoginFlow(w http.ResponseWriter, r *http.Request,
 			q.Set("refresh", "true")
 			r.URL.RawQuery = q.Encode()
 		}
+
+		// on OAuth2 flows, we need to use the RequestURL
+		// as the ReturnTo URL.
+		// This is because a user might want to switch between
+		// different flows, such as login to registration and login to recovery.
+		// After completing a complex flow, such as recovery, we want the user
+		// to be redirected back to the original flow.
+		if hlr.RequestUrl != "" && h.d.Config().OAuth2ProviderReturnToEnabled(r.Context()) {
+			// replace the return_to query parameter
+			q := r.URL.Query()
+			q.Set("return_to", hlr.RequestUrl)
```

Comment:
I'm setting the URL query parameter here instead of using passing in the URL to the `h.NewLoginFlow` function since we do not persist the return_to URL. So when retrieving the login flow later with the same ID will not give us the `return_to` URL since it is computed from the request URL at runtime. 

This was the cleanest approach I could find.